File: Chapter-5-Classes-and-Objects/011-a-sample-program-to-demonstrate-how-friend-functions-works-as-bridge-between-the-classes.cpp
// a sample program to demonstrate how friend functions works as bridge between the classes
#include <bits/stdc++.h>

using namespace std;

class ABC; /// forward declaration

class XYZ
{
    int x;

public:
    void setValue(int value)
    {
        x = value;
    }
    friend void maxValue(XYZ, ABC);
};

class ABC /// forward declaration
{
    int a;

public:
    void setValue(int value)
    {
        a = value;
    }
    friend void maxValue(XYZ, ABC);
};

void maxValue(XYZ m, ABC n) /// definition of friend
{
    if (m.x > n.a)
        cout << "The max value is : " << m.x << endl;
    else
        cout << "The max value is : " << n.a << endl;
}

int main()
{
    ABC abc;
    abc.setValue(10);

    XYZ xyz;
    xyz.setValue(20);

    maxValue(xyz, abc);
}

// output
// The max value is : 20

File: Chapter-5-Classes-and-Objects/011-a-sample-program-to-demonstrate-how-friend-functions-works-as-bridge-between-the-classes.py
class XYZ:
    def __init__(self):
        self.x = 0

    def set_value(self, value):
        self.x = value

class ABC:
    def __init__(self):
        self.a = 0

    def set_value(self, value):
        self.a = value

def max_value(m, n):
    if m.x > n.a:
        print("The max value is ", m.x)
    else:
        print("The max value is ", n.a)

if __name__ == '__main__':
    abc = ABC()
    abc.set_value(10)

    xyz = XYZ()
    xyz.set_value(20)

    max_value(xyz, abc)
